lib/class-wp-block-templates-registry.php

```diff
@@ -0,0 +1,272 @@
+		 * @param array                    $args          Optional. Array of block template arguments.
+		 * @return WP_Block_Template|false The registered block template on success, or false on failure.
+		 */
+		public function register( $template_name, $args = array() ) {
+
+			$template = null;
+			if ( $template_name instanceof WP_Block_Template ) {
+				$template      = $template_name;
+				$template_name = $template->name;
+			}
+
+			if ( ! is_string( $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block template names must be a string.', 'gutenberg' ),
+					'6.7.0'
+				);
+				return new WP_Error( 'template_name_no_string', __( 'Block template names must be a string.', 'gutenberg' ) );
+			}
+
+			if ( preg_match( '/[A-Z]+/', $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block template names must not contain uppercase characters.', 'gutenberg' ),
+					'6.7.0'
+				);
+				return new WP_Error( 'template_name_no_uppercase', __( 'Block template names must not contain uppercase characters.', 'gutenberg' ) );
+			}
+
+			$name_matcher = '/^[a-z0-9-]+\/\/[a-z0-9-]+$/';
+			if ( ! preg_match( $name_matcher, $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block template names must contain a namespace prefix. Example: my-plugin//my-custom-template', 'gutenberg' ),
+					'6.7.0'
+				);
+				return new WP_Error( 'template_no_prefix', __( 'Block template names must contain a namespace prefix. Example: my-plugin//my-custom-template', 'gutenberg' ) );
+			}
+
+			if ( $this->is_registered( $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					/* translators: %s: Template name. */
+					sprintf( __( 'Template "%s" is already registered.', 'gutenberg' ), $template_name ),
+					'6.7.0'
+				);
+				/* translators: %s: Template name. */
+				return new WP_Error( 'template_already_registered', __( 'Template "%s" is already registered.', 'gutenberg' ) );
+			}
+
+			if ( ! $template ) {
+				$theme_name = get_stylesheet();
+				$slug       = isset( $args['slug'] ) ? $args['slug'] : explode( '//', $template_name )[1];
+
+				$template                 = new WP_Block_Template();
+				$template->id             = $theme_name . '//' . $slug;
+				$template->theme          = $theme_name;
+				$template->plugin         = isset( $args['plugin'] ) ? $args['plugin'] : '';
+				$template->author         = null;
+				$template->content        = isset( $args['content'] ) ? $args['content'] : '';
+				$template->source         = 'plugin';
+				$template->slug           = $slug;
+				$template->type           = 'wp_template';
+				$template->title          = isset( $args['title'] ) ? $args['title'] : '';
+				$template->description    = isset( $args['description'] ) ? $args['description'] : '';
+				$template->status         = 'publish';
+				$template->has_theme_file = true;
+				$template->origin         = 'plugin';
+				$template->is_custom      = true;
```

Comment:
Why is this "true"?

Reply:
I might be wrong, but according to https://github.com/WordPress/gutenberg/pull/36911, my understanding is that `is_custom` means "it's not a default template from WP". In other words, any theme or plugin template should have `is_custom` set to true. Please correct me if I'm wrong.

Reply:
Now the question: is can a plugin override or register a default WP template (like archive.html or index.html or something).

Reply:
> Now the question is: Can a plugin override or register a default WP template (like archive.html or index.html or something).

Theme templates [have priority](https://github.com/WordPress/gutenberg/pull/61577/files#diff-022cdd9afc1bedeb4dd66327c7bfcd4048dd87cc16cc809e6234688fc589db7eR37-R52) over plugin-registered templates. However, if a theme doesn't have a template, the plugin can register one with that slug. Ie: a plugin could register a `category` template that would be used in the [template hierarchy](https://i0.wp.com/developer.wordpress.org/files/2014/10/Screenshot-2019-01-23-00.20.04.png?ssl=1), but it the theme had a template with the same name, the theme template would take priority.

Does it make sense or should we block registering templates with certain slugs?

Reply:
So there are two things here:

 - If we allow plugins to register "default WP templates", I feel the order of priority should be: "User > Plugin > Theme". So User (saved templates) should be used otherwise plugin otherwise theme templates.
 - If we allow plugins to register "default WP templates", `is_custom` property need to be set accordingly (depending on the slug I guess)

Reply:
I'm also ok starting without support for WP default templates in this API but I think we should ultimately support it.

Reply:
Similarly we need to investigate flows like:
1. theme hasn't a template (ex 2024 with author archive `author`)
2. user creates that template
3. then plugin registers `author`
4. interactions

Flow we need to check are user created templates in conjunction with the API.

I have a small video that shows some nuances.. 

https://github.com/user-attachments/assets/3667fbab-cf3c-4c7e-9bcd-17fea4ac514e

In general it will be very useful to provide a couple of examples in the PR's description, how to properly test the new API.

Reply:
> I see your point, but I think this might conflict with what makes sense for 3rd-party templates, where I think "User > Theme > Plugin" is better. For example, if a plugin registers a Single Product template, themes should be able to override it easily.
> If a plugin really needs to have priority over theme templates, it can always hook into *_template_hierarchy. What do you think?

To be honest I'm not sure and I'm second guessing myself after your reply :P. I feel this is a place where we should look at what we may be doing elsewhere for other things that can be defined both in themes, plugins and edited by users. Is this the first "entity" that supports this? I can be on board with any order but I would prefer if we stay consistent regardless of the "entity".

> You're right, in my original PR I didn't account for the overlap between plugin-registered templates and default/user-created templates. I will be taking a look at those and will update the testing steps.

Am I right in assuming that if there's an overlap, the REST API will only return one item per slug (like we do today for theme + user)

Reply:
>Ok, so if I'm understanding it right, for now we would block registering templates with [these slugs](https://github.com/WordPress/wordpress-develop/blob/3296c6b842adc5e7c8c78cd06db9ccca2ee56ce6/src/wp-includes/block-template-utils.php#L142-L207), correct?

The `is_custom` flag is for any template that is part of WP template hierarchy, which can also be created from the `add template` menu through the UI. That also means templates like `author:admin` are not considered custom. I'm mentioning this because the link shared for the slugs includes only the default templates.

Reply:
> What do you think?

I think it makes sense indeed. so let's roll with that.

> Yes, currently, in the way it's implemented in this PR, it depends on the order templates are registered. Any suggestions on how it should behave otherwise? I'm a bit wary of throwing errors because if we officially support plugins overriding default templates, we should support more than one plugin trying to register the same template, no?

Yes, if we're just comparing plugin registration, either first or last registered works for me. 

> Ok, I'm getting a bit confused by this. I tested and is_custom is false when Author: Admin is created from the UI but it's true when author-admin.html is provided from the theme. Is that expected? It feels a bit counter-intuitive to me. 😅 I would keep plugin-registered templates consistent with theme templates over user-created templates, no?

I think for me the "user" ones are kind of broken here and we should rely on the "theme" behavior.

Reply:
> Similarly we need to investigate flows like:

Heads-up that I fixed this specific flow in my latest commit (046dbec037ecdc4ca73d09340f2bf26ab18deaaa). The `origin` property wasn't set properly for plugin-registered templates edited by the user. So at the time of trying to reset them, it wasn't sending the correct `origin` property.

> In general it will be very useful to provide a couple of examples in the PR's description, how to properly test the new API.

I added two new test flows which should cover what we discussed in this thread: _Registering WP default templates_ and _Registering templates already customized by the user_.